lib/http3/server.c

```diff
@@ -1195,6 +1273,91 @@ static void do_send_informational(h2o_ostream_t *_ostr, h2o_req_t *_req)
+{
+    struct st_h2o_http3_server_stream_t *stream = _tunnel->data;
+
+    stream->proceed_requested = 0;
+
+    /* append DATA frame */
+    if (len != 0) {
+        h2o_vector_reserve(&stream->req.pool, &stream->sendbuf.vecs, stream->sendbuf.vecs.size + 2);
+        /* DATA frame header */
+        size_t header_size = flatten_data_frame_header(stream, stream->sendbuf.vecs.entries + stream->sendbuf.vecs.size++, len);
+        /* payload */
+        struct st_h2o_http3_server_sendvec_t *vec = stream->sendbuf.vecs.entries + stream->sendbuf.vecs.size++;
+        h2o_sendvec_init_raw(&vec->vec, bytes, len);
+        vec->entity_offset = stream->sendbuf.final_body_size;
+        stream->sendbuf.final_body_size += len;
+        /* update final offset */
+        stream->sendbuf.final_size += header_size + len;
+    }
+
+    /* EOS */
+    if (err != NULL)
+        shutdown_response(stream);
+
+    finalize_do_send(stream);
+}
+
+void tunnel_write(struct st_h2o_http3_server_stream_t *stream)
+{
+    size_t bytes_to_send;
+
+    assert(!stream->tunnel->up.is_inflight);
+
+    if ((bytes_to_send = stream->req_body->size) == 0) {
+        if (quicly_recvstate_transfer_complete(&stream->quic->recvstate)) {
+            stream->tunnel->tunnel->destroy(stream->tunnel->tunnel);
+            stream->tunnel->tunnel = NULL;
+            shutdown_response(stream);
+        }
+        return;
+    }
+
+    /* move chunk of data into stream->tunnel.up.buf */
+    if (bytes_to_send > sizeof(stream->tunnel->up.bytes_inflight))
+        bytes_to_send = sizeof(stream->tunnel->up.bytes_inflight);
+    memcpy(stream->tunnel->up.bytes_inflight, stream->req_body->bytes, bytes_to_send);
+    stream->tunnel->up.is_inflight = 1;
+    h2o_buffer_consume(&stream->req_body, bytes_to_send);
+
+    /* send */
+    stream->tunnel->tunnel->write_(stream->tunnel->tunnel, stream->tunnel->up.bytes_inflight, bytes_to_send);
+}
+
+void tunnel_write_delayed(h2o_timer_t *timer)
+{
+    struct st_h2o_http3_server_stream_t *stream =
+        H2O_STRUCT_FROM_MEMBER(struct st_h2o_http3_server_tunnel_t, up.delayed_write, timer)->stream;
+    tunnel_write(stream);
+}
+
+static void tunnel_on_write_complete(h2o_httpclient_tunnel_t *tunnel, const char *err)
+{
+    struct st_h2o_http3_server_stream_t *stream = tunnel->data;
+
+    assert(stream->tunnel->up.is_inflight);
+    stream->tunnel->up.is_inflight = 0;
+
+    tunnel_write(stream);
+}
+
+static void establish_tunnel(h2o_req_t *req, h2o_httpclient_tunnel_t *tunnel, uint64_t idle_timeout)
+{
+    struct st_h2o_http3_server_stream_t *stream = H2O_STRUCT_FROM_MEMBER(struct st_h2o_http3_server_stream_t, req, req);
+
+    stream->tunnel->tunnel = tunnel;
+    tunnel->data = stream;
+    tunnel->on_write_complete = tunnel_on_write_complete;
+    tunnel->on_read = tunnel_on_read;
+
+    write_response(stream);
+    set_state(stream, H2O_HTTP3_SERVER_STREAM_STATE_SEND_BODY);
+
+    finalize_do_send(stream);
```

Comment:
In case of a tunnel, after the tunnel is established, the tunnel should first invoke the `on_read` callback then the application should call `proceed_read`. I the invocation of `finalize_do_send` is violates that contract.

Or, do we want to pass the first chunk as the argument of `establish_tunnel`? Doing so provides us better alignment to the `do_send` abstraction that we have.

Reply:
Fixed in 6d75dfc7a.